### PR TITLE
HTTPEndpoint: role for HTTP endpoint on a channel or reactor

### PR DESCRIPTION
**Pull request by robn**
Adds a role that automatically hooks up an endpoint to a channel or reactor, including authentication if the config demands it.

The base HTTPServer remains available for anything that wants to do something special; in that case, they're on their own and will have to handle auth and anything else directly.

I've changed HTTPServer to call the apps with the PSGI env, not Plack::Request. That's the PSGI way anyway, but also makes the wrappers much simpler because we don't need to keep unwrapping and rewrapping the request for things that expect an env.

I think I like the feel of this. Anything you might want is in one place, common config for all channels and reactors that might want a HTTP endpoint, and its pretty easy to set up.

There might be some structural problems; I'm not sure about the `http_app` method really, which feels a bit off. Maybe it wants to return an app instead of being the app? I'm also still not used to roles, so they might be a bit oddly assembled.

I think I like this better than #78.

### CODE REVIEW

File: t/httpendpoint.t [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use IO::Async::Test;
+use IO::Async::Timer::Periodic;
+use Net::Async::HTTP;
+use Plack::Response;
+use MIME::Base64 'encode_base64';
+use Synergy::Hub;
+
+package Synergy::Channel::Test::HTTPEndpoint {
+
+use Moose;
```
[the remark this answers is not in the document]
Comment (robn) [reply t/httpendpoint.t]: I guess if I wasn't sure, and you pointed it out, that's enough to beat the ">= 1 people have an opinion" threshold that I just made up. Will do!

**Comment by robn** [comment]
Re-pushed with the actual role package, rookie 🙄

File: lib/Synergy/Channel/Twilio.pm [hunk 2]
```diff
@@ -25,86 +26,84 @@ has numbers => (
...
-        [ 'Content-Type', 'application/json' ],
-        [ "{}\n" ],
-      ];
-    }
+  my $param = $req->parameters;
+  my $from  = $param->{From} // '';
 
-    my $who = $self->hub->user_directory->user_by_channel_and_address(
-      $self,
+  unless (($param->{AccountSid}//'') eq $self->sid) {
```
[the remark this answers is not in the document]
Comment (wolfsage) [reply lib/Synergy/Channel/Twilio.pm]: Fair

File: lib/Synergy/Role/HTTPEndpoint.pm [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      )->wrap($app);
+    }
+
+    return $app;
+  },
+);
+
+sub BUILD($self, @) {
+  confess "http_username and http_password must be supplied together"
+    if $self->has_http_username ^ $self->has_http_password;
```
Comment (wolfsage) [inline lib/Synergy/Role/HTTPEndpoint.pm]: Oh man, I really gotta use `^` more

**Pull request merged**